tracks/julia/exercises/raindrops/mentoring.md

```diff
@@ -186,26 +206,56 @@ function raindrops4(number)
 end
 ```
 
-Try changing `s = "Pling"` to `s *= "Pling"` and benchmarking with 3, 14, and 15 to see what's going on.
+We can continue this idea of reducing concatenations to further improve performance.
+
+```julia
+function raindrops6(n)
+    (f3 = n % 3 == 0) && (s = "Pling")
+    (f5 = n % 5 == 0) && (s = f3 ? "PlingPlang" : "Plang")
+    # To replace all concatenations,
+    # the next line would have to contain another layer of branching
+    # At that point you might as well use the full if tree solution
+    (f7 = n % 7 == 0) && (s = f3 | f5 ? string(s, "Plong") : "Plong")
+    return f3 | f5 | f7 ? s : string(n)
+end
+```
+
+This solution retains the structure and compactness of the previous solutions, but achieves very
+good performance (similar to switch) because string concatenation is only required for numbers divisble by 3 and/or 5
+and 7. 
```

Comment:
The end of this sentence could be clearer. I'd also say that this solution is quite a lot harder to understand than the other compact one. Changing the tests to test and assign and using lots of ternary expressions means there's a lot going on.

Reply:
What about saying that it retains the structure, sacrifices some readability because of the added complexity but also improves performance to where this solution is as fast as the switch? We can also mention that it will depend on the situation if improving performance at the cost of readability is actually advantageous. Depending on the need for performance you may choose different variants.

Reply:
I don't think it really does retain much of the structure. It has the same number of lines, but I don't think we should be encouraging assignments on the left hand side of an expression, it's already a little frowned on to be doing it on the right hand side of a short-circuiting op.

Here's the solution re-written to be a bit easier to read. It also makes clear that this is basically just the if/else tree solution:

```julia
function raindrops6(n)
    f3 = n % 3 == 0
    f5 = n % 5 == 0
    f7 = n % 7 == 0

    f3 && (s = "Pling")
    f5 && (s = f3 ? "PlingPlang" : "Plang")
    f7 && (s = f3 || f5 ? s * "Plong" : "Plong")

    # If we wanted to avoid all concatenations, we could use this instead, but
    # at that point, this solution becomes equivalent to the tree of if/elses
    # f7 && (s = f3 ? (f5 ? "PlingPlangPlong" : "PlingPlong")
    #               : (f5 ? "PlangPlong" : "Plong"))

    return f3 || f5 || f7 ? s : string(n)
end
```

I'm just not convinced that this solution offers much extra learning over the compact solution, and, if we do include it, I think it makes more sense to make it equivalent to the if/else tree to show that that solution can be written concisely, maybe replacing the "switch" solution (which is just a worse version of the lookup solution, IMO)

Again, if others prefer how this PR already is, then that's fine. This is not a veto :)

Reply:
I like the code you posted there. Spacing it out and showing how you can make it more and more like the if tree (and thereby also improve performance) by adding comments is a more concise and clear way to say the same thing.

Reply:
I guess I shouldn't say "equivalent", because the conditions are in a different order to the tree. It's just similar.

Reply:
OTOH, if you replace just the bottom level of the if tree with ternary expressions you get this fairly short thing:

```jl
function raindrops(n)
    f3 = n % 3 == 0
    f5 = n % 5 == 0
    f7 = n % 7 == 0

    if f3
        if f5
            f7 ? "PlingPlangPlong" : "PlingPlang"
        else
            f7 ? "PlingPlong" : "Pling"
        end
    else
        if f5
            f7 ? "PlangPlong" : "Plong"
        else
            f7 ? "Plong" : string(n)
        end
    end
end
```

Or, golfed:

```jl
function raindrops(n)
    f3 = n % 3 == 0
    f5 = n % 5 == 0
    f7 = n % 7 == 0
    f3 ? f5 ? f7 ? "PlingPlangPlong" : "PlingPlang"
            : f7 ? "PlingPlong"      : "Pling"
       : f5 ? f7 ? "PlangPlong"      : "Plang"
            : f7 ? "Plong"           : string(n)
end
```

And that's just better in the number of comparisons done than raindrops6 (this one is always three comparisons, raindrops6 is 4 in the best case and 6 in the worst).